Unsafe lint will also check for unsafe functions, traits, and implementations

ipetkov opened the pull request:
This allows warning or forbidding all uses of unsafe code, whereas
previously only unsafe blocks were caught by the lint.

The lint has been renamed from `unsafe-blocks` to `unsafe-code` to
reflect its new purpose.

This is a minor [breaking-change]

Closes #22430

rust-highfive commented:
r? @huonw

(rust_highfive has picked a reviewer for you, use r? to override)

kmcallister commented on src/librustc/lint/builtin.rs:
```diff
@@ -1269,27 +1269,40 @@ impl LintPass for UnusedUnsafe {
+    UNSAFE_CODE,
     Allow,
-    "usage of an `unsafe` block"
+    "usage of an `unsafe` code"
```
nit: drop the article `an`

ipetkov replied:
oops, totally missed that, thanks for the catch!

ipetkov commented:
@huonw I added tests for (both declaring and implementing) unsafe methods of traits. My previous implementation of the lint didn't catch these, so I refactored the code to catch them as well!

huonw commented:
@bors r+ be0d

Thanks!

Manishearth commented:
Needs https://github.com/Manishearth/rust/commit/bf13792b319ea0f85900fba134f4102037369f47 to work.

ipetkov commented:
Thank you for the fix @Manishearth!
